File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/AppBasicDataService.java
package sk.styk.martin.apkanalyzer.business.service;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import sk.styk.martin.apkanalyzer.model.detail.AppSource;
import sk.styk.martin.apkanalyzer.model.list.AppListData;
import sk.styk.martin.apkanalyzer.util.AppBasicInfoComparator;

/**
 * Retrieve apps installed on device
 * <p>
 * @author Martin Styk
 * @version 14.06.2017.
 */
public class AppBasicDataService {

    private PackageManager packageManager;

    public AppBasicDataService(@NonNull PackageManager packageManager) {
        this.packageManager = packageManager;
    }

    public AppBasicDataService(@NonNull Context context) {
        this.packageManager = context.getPackageManager();
    }

    @NonNull
    public List<AppListData> getAll() {

        List<PackageInfo> applications = packageManager.getInstalledPackages(0);

        List<AppListData> packages = new ArrayList<>(applications.size());

        for (PackageInfo packageInfo : applications) {
            if (packageInfo.applicationInfo != null)
                packages.add(new AppListData(packageInfo, packageManager));
        }

        Collections.sort(packages, AppBasicInfoComparator.INSTANCE);

        return packages;
    }

    @NonNull
    public List<AppListData> getForSources(boolean allowSystem, @NonNull AppSource... appSources) {

        List<AppListData> appListData = getAll();
        List<AppListData> results = new ArrayList<>();
        List<AppSource> appSourceList = Arrays.asList(appSources);

        for (AppListData data : appListData) {
            if (appSourceList.contains(data.getSource()) && (allowSystem || !data.isSystemApp())) {
                results.add(data);
            }
        }

        return results;
    }

    @NonNull
    public List<String> getAllPackageNames() {

        List<ApplicationInfo> applications = packageManager.getInstalledApplications(PackageManager.GET_META_DATA);

        List<String> packages = new ArrayList<>(applications.size());

        for (ApplicationInfo applicationInfo : applications) {
            packages.add(applicationInfo.packageName);
        }

        Collections.sort(packages);

        return packages;
    }

    @NonNull
    public List<AppListData> getForPackageNames(@NonNull List<String> packageNames) {

        List<AppListData> packages = new ArrayList<>(packageNames.size());

        for (String packageName : packageNames) {
            PackageInfo packageInfo = null;
            try {
                packageInfo = packageManager.getPackageInfo(packageName, 0);
            } catch (PackageManager.NameNotFoundException e) {
                continue;
            }

            if (packageInfo.applicationInfo != null)
                packages.add(new AppListData(packageInfo, packageManager));
        }

        Collections.sort(packages, AppBasicInfoComparator.INSTANCE);

        return packages;
    }


}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/analysis/logic/launcher/AppBasicDataService.kt
package sk.styk.martin.apkanalyzer.business.analysis.logic.launcher

import android.content.pm.PackageManager
import android.support.annotation.WorkerThread
import sk.styk.martin.apkanalyzer.model.detail.AppSource
import sk.styk.martin.apkanalyzer.model.list.AppListData
import sk.styk.martin.apkanalyzer.util.AppBasicInfoComparator
import java.util.*

/**
 * Retrieve apps installed on device
 *
 *
 * @author Martin Styk
 * @version 14.06.2017.
 */
@WorkerThread
class AppBasicDataService(val packageManager: PackageManager) {

    fun getAllPackageNames(): List<String> {
        val applications = packageManager.getInstalledApplications(PackageManager.GET_META_DATA)
        return applications
                .asSequence().mapTo(ArrayList<String>(applications.size)) { it.packageName }
                .asSequence().sorted().toList()
    }

    fun getAll(): List<AppListData> {
        val applications = packageManager.getInstalledPackages(0)
        val packages = ArrayList<AppListData>(applications.size)

        applications.forEach {
            if (it.applicationInfo != null)
                packages.add(AppListData(it, packageManager))
        }

        return packages.sortedWith(AppBasicInfoComparator.INSTANCE)
    }

    fun getForSources(allowSystem: Boolean, vararg appSources: AppSource): List<AppListData> {
        val appSourceList = Arrays.asList(*appSources)
        return getAll().filter { appSourceList.contains(it.source) && (allowSystem || !it.isSystemApp) }
    }

    fun getForPackageNames(packageNames: List<String>): List<AppListData> {

        val packages = ArrayList<AppListData>(packageNames.size)

        packageNames.forEach {
            try {
                val packageInfo = packageManager.getPackageInfo(it, 0)
                if (packageInfo?.applicationInfo != null)
                    packages.add(AppListData(packageInfo, packageManager))

            } catch (e: PackageManager.NameNotFoundException) {
                // continue
            }
        }

        return packages.sortedWith(AppBasicInfoComparator.INSTANCE)
    }
}
